// evergreen-ui/commonjs/tag-input/src/Tag.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _badges = require("../../badges");

var _icon = require("../../icon");

var _scales = require("../../scales");

/**
 * @overview TagInput accepts multiple values that can be individually removed
 */
var Tag =
/*#__PURE__*/
function (_React$PureComponent) {
  (0, _inherits2.default)(Tag, _React$PureComponent);

  function Tag() {
    (0, _classCallCheck2.default)(this, Tag);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Tag).apply(this, arguments));
  }

  (0, _createClass2.default)(Tag, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          children = _this$props.children,
          onRemove = _this$props.onRemove,
          isRemovable = _this$props.isRemovable,
          props = (0, _objectWithoutProperties2.default)(_this$props, ["children", "onRemove", "isRemovable"]);
      var badgeStyles = {
        alignItems: 'center',
        display: 'inline-flex',
        fontWeight: 400,
        textTransform: 'none'
      };

      if (isRemovable) {
        badgeStyles.paddingRight = (0, _scales.minorScale)(1);
      }

      return _react.default.createElement(_badges.Badge, (0, _extends2.default)({
        isInteractive: true
      }, badgeStyles, props), children, isRemovable && _react.default.createElement(_icon.Icon, {
        icon: "cross",
        marginLeft: (0, _scales.minorScale)(1),
        onClick: onRemove,
        size: (0, _scales.minorScale)(3)
      }));
    }
  }]);
  return Tag;
}(_react.default.PureComponent);

Tag.displayName = "Tag";
(0, _defineProperty2.default)(Tag, "propTypes", {
  /** The badge content */
  children: _propTypes.default.node,

  /**
   * Callback invoked when the removal icon is clicked.
   * (event) => void
   */
  onRemove: _propTypes.default.func,

  /** Whether or not the tag can be removed. */
  isRemovable: _propTypes.default.bool
});
var _default = Tag;
exports.default = _default;
